### ISSUE TITLE
A wrong pre-trained weights?

### ISSUE THREAD

**Issue by Fjr9516** [voxelmorph/voxelmorph#450]
**Task** (what are you trying to do/register?)

I am trying to create deformable templates by using the pre-trained weights you provided (conditional atlas). And I found the outputs of layer atlas_gen always give me noisy maps, it is a simple test, so I guess maybe you put the wrong one?

**What have you tried**

Using this model: Creation of Deformable Templates. 

**Details of experiments**
key code:
```
atlas_gen_model = tf.keras.models.Model(model.inputs[:1], model.get_layer('atlas_gen').output)
input_samples = [tf.keras.utils.to_categorical(np.arange(pheno_shape), pheno_shape)]
pred = atlas_gen_model.predict(input_samples)
```

**Comment by kvttt** [comment]
Hi @Fjr9516, I don't have the exact same issue but I find it weird that the atlas generation model actually adds the rough template to the decoded atlas at the very last step before outputting. The issue I am facing right now is that the atlas generation model seems to ignore the decoded attributes and produces atlases that look just like the rough template.

**Comment by Fjr9516** [comment]
> Hi @Fjr9516, I don't have the exact same issue but I find it weird that the atlas generation model actually adds the rough template to the decoded atlas at the very last step before outputting. The issue I am facing right now is that the atlas generation model seems to ignore the decoded attributes and produces atlases that look just like the rough template.

Hi, I think we are facing the same issue. I showed the outputs from "atlas_gen" directly, while you showed the outputs of the "atlas" layer. The only difference is whether the rough template is added to the decoded atlas. If you are also trying to use the provided pre-trained weights, my guess is they provided the wrong one.

**Comment by adalca** [comment]
hmm, did either/both of you use the code in the colab? It might be helpful to try to debug/figure this out with something simple like MNIST.

@kvttt the reason to we add the initial rough template is that it acts like an "initialization", in a sense having the atlas part of the network learn just the residual.

**Comment by Fjr9516** [comment]
> hmm, did either/both of you use the code in the colab? It might be helpful to try to debug/figure this out with something simple like MNIST.
> 
> @kvttt the reason to we add the initial rough template is that it acts like an "initialization", in a sense having the atlas part of the network learn just the residual.

Hi, thanks for your reply! I used the code in the GitHub repo and wrote test code by myself. And the pre-trained weights file was downloaded via this link: http://people.csail.mit.edu/adalca/voxelmorph/atlas_creation_cond_NCC_1022.h5. Actually, I have tried the MNIST demo you put in the colab and successfully tried MNIST model. Thanks for providing that, it is useful. But it should be more useful to use your pre-trained weights since my interest is the brain. But when I used the Model in h5 format weights you provided (conditional atlas), the outputs from the "atlas_gen" layer only give me noisy maps no matter the input attributes.  

The test code is shown here:
```
        pheno_shape = 14 # number of attributes, 7 ages for male and 7 for female
        inshape = (160,  192, 224)  # (160, 192, 224)
        nfeats = 1
        # build the model
        model = vxm.networks.ConditionalTemplateCreation(
            inshape,
            pheno_input_shape=pheno_shape,
            nb_unet_features=[enc_nf, dec_nf],
            conv_nb_features=4,
            conv_nb_levels=0,
            extra_conv_layers=3,
            src_feats=nfeats,
            trg_feats=nfeats
        )
 
        # load pre-trained weights
        model_weights = os.path.abspath("../weights/atlas_creation_cond_NCC_1022.h5")
        model.load_weights(model_weights, by_name=True, skip_mismatch=True)
 
        # extract atlas_model
        atlas_gen_model = tf.keras.models.Model(model.inputs[:1], model.get_layer('atlas_gen').output)

        input_samples = [tf.keras.utils.to_categorical(np.arange(pheno_shape), pheno_shape)]
        pred = atlas_gen_model.predict(input_samples)
```
And the outputs look like this:
![1657121351085](https://user-images.githubusercontent.com/31833999/177587644-740d64c7-87bc-47bf-aa2b-e0d664280395.png)
Any suggestion would be grateful!

**Comment by adalca** [comment]
Oh I wonder if the code has changed substantially since we made those 
weights and that's why. You have "skip_mismatch=True", I'm assuming that 
means that it doesn't load weights in certain scenarios? SO perhaps it's 
not loading the weights?

**Comment by kvttt** [comment]
Hi @Fjr9516, I think I'm doing the same task as yours except that I am training from scratch on my own data. I have run 200 epochs so far. The losses seem to converge and the atlas I'm getting looks similar to the rough template from average. But as @adalca mentioned, I think the model did learned the residual. Attached are the residual between the final output and rough template, the rough template, and the final atlas.

![image](https://user-images.githubusercontent.com/99367900/177643922-73656c80-9a3e-45ee-861a-bb293046c68a.png)

![image](https://user-images.githubusercontent.com/99367900/177643940-1466fc07-74aa-4a39-9b08-9621283fecf3.png)

![image](https://user-images.githubusercontent.com/99367900/177643952-cea96cfa-898c-4cb0-a933-8a9fb09e5581.png)

With that being said, I think your problem probably has to do with parameters not properly loaded, because the model seems to be working as expected when I am training from scratch. 

And by the way, @adalca thanks for the clarification!

**Comment by adalca** [comment]
@kvttt just so i understand, the last image you showed is the final template when trained from scrach? That certainly seems much blurrier than I would expect.

**Comment by kvttt** [comment]
> @kvttt just so i understand, the last image you showed is the final template when trained from scrach? That certainly seems much blurrier than I would expect. 

Yes. And to provide more details, I only trained for around 200 epochs to get this template. And previously, I have training the rest of the network (the UNet portion) separately as a registration network. And I am currently fixing the parameters of the registration network and tuning the parameters of the conditional template creation network. In addition, I am testing the effect of another loss on the result and I am simultaneously running a control experiment without that loss (i.e. using only similarity loss for this step). Finally, I am implementing all this using PyTorch. So, there is a lot of stuff that could go wrong. I will update on how it goes after I finished training.

**Comment by Fjr9516** [comment]
Hi @kvttt and @adalca, thanks for your replies! I think you are all right, the load_weights() function didn't work properly. It's caused by different settings between the trained model and the source code in the repo, where I previously assumed they were the same. Based on my investigation of the provided weights, I found that your model sets the input attribute size at 2. But according to the paper, you said "For conditional template, we do this for 7 ages equally spaced between 15 and 90 years old, for both males and females". Therefore, you should have seven attributes. Could you elaborate on how these two pheno sizes can represent seven different attributes? Any suggestion would be grateful!

**Comment by kvttt** [comment]
> Hi @kvttt and @adalca, thanks for your replies! I think you are all right, the load_weights() function didn't work properly. It's caused by different settings between the trained model and the source code in the repo, where I previously assumed they were the same. Based on my investigation of the provided weights, I found that your model sets the input attribute size at 2. But according to the paper, you said "For conditional template, we do this for 7 ages equally spaced between 15 and 90 years old, for both males and females". Therefore, you should have seven attributes. Could you elaborate on how these two pheno sizes can represent seven different attributes? Any suggestion would be grateful!

I would love to know the answer to this question too. Currently I am treating the attribute age as a continuous variable. But it sounds like treating age as a categorical variable and using one-hot encoding is also viable.

**Comment by adalca** [comment]
That quote from the paper is just for evaluation -- the model is continuous (for age)
To answer your question, we have 2 inputs for the 2 attributes: age and sex. Age is continuous (I think 0-1 if we normalized right, it's been a long time!) and sex was 0/1. So nothign was one-hot encoded.

**Comment by Fjr9516** [comment]
Hi, @adalca, thanks for clarifying! I tested the model with both non-normalized and normalized age attributes, and the non-normalized one performed well. So I guess it wasn't normalized :). I believe @kvttt mentioned one-hot encoding because he also went through the colab demo on the MNIST dataset, where you use one-hot encoding as the input attribute. I also used one-hot encoding all the time to train my model on the brain data, but it's very hard to appreciate the differences among templates. Will try the continuous attribute. Thanks again!

**Comment by kvttt** [comment]
Hi @Fjr9516, similar to what you mentioned in your comment, I also compared using normalized and non-normalized age attributes. And in my case, the normalized version converges faster. Attached is the comparison:

<img width="380" alt="Screen Shot 2022-07-08 at 3 43 51 AM" src="https://user-images.githubusercontent.com/99367900/177977537-4d454be5-ee7a-418f-b94f-c757f2df35a0.png">

I observed that (in my case) the non-normalized version converges to a bigger value for the similarity loss. But @Fjr9516, I'm not mentioning this to invalidate your observation because I am currently training the atlas generation network with the registration network fixed and I am using only the similarity loss during training. And I think it could be possible that this is the cause of the discrepancy.

**Comment by adalca** [comment]
> Hi, @adalca, thanks for clarifying! I tested the model with both non-normalized and normalized age attributes, and the non-normalized one performed well. So I guess it wasn't normalized :). I believe @kvttt mentioned one-hot encoding because he also went through the colab demo on the MNIST dataset, where you use one-hot encoding as the input attribute. I also used one-hot encoding all the time to train my model on the brain data, but it's very hard to appreciate the differences among templates. Will try the continuous attribute. Thanks again!

Makes sense -- the demo had a one-hot because we were working with 10 classes I think, rather than continuous or binary.

> Hi @Fjr9516, similar to what you mentioned in your comment, I also compared using normalized and non-normalized age attributes. And in my case, the normalized version converges faster. Attached is the comparison:

We usually recommend normalization because of stability when training from scratch. @Fjr9516 Did you also mean that you trained from scratch in these two modes?

**Comment by Fjr9516** [comment]
Hi @kvttt, nice comparison experiment! I'd love to see it! But what I mentioned was just the test using the pre-trained model. So the pre-trained model was trained by using the non-normalized age attribute according to my investigation. Here are the outputs of learned templates when using 0, and 1 as the age attribute, respectively (the sex attribute is set to 1 all the time):
![0](https://user-images.githubusercontent.com/31833999/178051925-869ea6e5-ba24-4b6e-b0ab-a9b997d99d28.png)
![1](https://user-images.githubusercontent.com/31833999/178051932-00ee4946-5664-43e9-aea5-aa05f3774bbc.png)

And here are the learned templates when using 15, and 90, respectively:
![15](https://user-images.githubusercontent.com/31833999/178051959-4bf1d30f-8115-495b-aa52-0ff513c78218.png)
![90](https://user-images.githubusercontent.com/31833999/178051954-10b3beb9-c990-4d9f-8cb4-dde7ac39bb6b.png)

> We usually recommend normalization because of stability when training from scratch. @Fjr9516 Did you also mean that you trained from scratch in these two modes?

Hi, @adalca, I only trained from scratch by using one-hot encoding so far. With regard to generalization, I have a question: if the training datasets don't contain images at ages 0~15, can the generator also learn from the rest of the data?

**Comment by adalca** [comment]
We didn't test something as dramatic as 0~15 if there is no data. We did see that there's good learning when you have very little data in some middle age range (e.g. if you have tons of 15-25, 40-90 but very few 25-40). 

We have a separate project on combining datasets across all ages (pediatric and adult) but the model is not out yet.

**Comment by Fjr9516** [comment]
> We didn't test something as dramatic as 0~15 if there is no data. We did see that there's good learning when you have very little data in some middle age range (e.g. if you have tons of 15-25, 40-90 but very few 25-40).
> 
> We have a separate project on combining datasets across all ages (pediatric and adult) but the model is not out yet.

Thanks for your prompt answer! Looking forward to your new work!
